spec/asyncapi.md

```diff
@@ -552,8 +479,39 @@ Field Name | Type | Description
 
 This object MAY be extended with [Specification Extensions](#specificationExtensions).
 
+##### Server Variable Object Example
 
+```json
+{
+  "servers": {
+    "production": {
+      "host": "{org}.mycompany.com",
+      "path": "/api/v1",
```

Comment:
`path` ? where is this coming from?

Reply:
We were having a "URL" and now we have a host. A host does not have the "path" part so we have to add it separately.

Reply:
The other option is to remove the `protocol` field and go with a URL as we do now. Just not sure how this would play with protocols like `kafka-secure` (e.g., `kafka-secure://host:port/path`) and `secure-mqtt` (e.g., `secure-mqtt://host:port/path`), which are not "real" protocols. In such a case, we should probably get rid of the "secure" variants (that would include `https` and `wss`) and allow the user to specify it in the Security Requirement Object.

Reply:
I'm wondering if we are actually fixing something, or making things a bit more complex. You add now `path` example that is a typical REST example for me. All the other examples I know (for example for Kafka) include a version in the topic name. In case of websocket it is completely fine to add path to channel name, but also if someone wants, can add it to `url`. The only use case for `path` that makes sense is HTTP, but then, if it is protocol specify, shouldn't it end up in the binding?

so basically my question is, do we really need to change `url` to `host` and `path` or, as majority of people in https://github.com/asyncapi/spec/issues/274 wrote that we basically need to make description and examples better?

Reply:
There's are two things here:

1. We often show examples where the `url` is "whatever.example.com". That's not a URL and that should fail because it doesn't have the protocol part of the URL (you know, `protocol://`).
2. We have a field called `protocol` to define the protocol we want to use but it's already in the URL (!) ☝️ So you can end up with `url: kafka://broker.mycompany.com` and `protocol: https` in the same server definition, which is incongruent.

So we either remove the need to put a protocol in the `url` field (which makes it a host instead) or we remove the `protocol` field.

> You add now path example that is a typical REST example for me.

I think you're mixing concepts here. A path is not about REST or not even HTTP. A path is a part of a URL. URLs are composed by `protocol://`, `username:password` (optional`), `hostname`, `port` (optional), `/path/to/resource` (optional), `?query=parameters` (optional), and `#fragment` (optional). These have nothing to do with REST or HTTP. I can have an AMQP broker at `amqp://broker.mycompany.com/public` and another one at `amqp://broker.mycompany.com/partners`. Depending on the path (`public` or `partners`), I'll reach one broker or another. This is actually a common pattern, BTW.

So what I'm doing here is making `path` optional (defaulting to `/` as it's assumed by the URL RFC). In the majority of cases, this path will not be needed but in some cases, it will. And yeah, especially in the case of HTTP and WS it's going to be needed but it's also frequent in AMQP and MQTT brokers.

---

So yeah, I took the approach to keep the `protocol` field and convert the `url` field to a `host` (hostname + port) one. Another approach is to remove `protocol` altogether as it's already in the URL field. A matter of preference, I guess.

> so basically my question is, do we really need to change url to host and path or, as majority of people in https://github.com/asyncapi/spec/issues/274 wrote that we basically need to make description and examples better?

We definitely have to improve the examples but this incongruency goes beyond showing examples or improving description.

Reply:
Well one thing is "concepts" and one thing is "just a simple association"

When I see
```
"host": "{org}.mycompany.com",
"path": "/api/v1",
"protocol": "https"
```

My brains says "REST". Then my brain also reminds me of AsyncAPI 1.0 and [baseTopic](https://github.com/asyncapi/spec/tree/1.2.0#baseTopicString). And then it tries to map it to concepts 😄 but yeah, enough with my brain 😄 

Can we then improve the example and not put `https` protocol but something that will not cause brains to fart?
For example your `amqp` example. I guess there are different patterns, as I thought people prefer to split `public/private` or `public/partners` by host as then they do different authentication, rate limits and stuff like that. But yeah, it is flexible and people can do it as they wish.

or best would be to have 2 examples, one with path and one without it:
- websocket from kraken provides different hosts (URL 😄 ) https://docs.kraken.com/websockets/#connectionDetails
- websocket from geminig (https://docs.gemini.com/websocket-api/#market-data) where host is `api.gemini.com` and `pathname` is `v1`

Reply:
also the origin of my question

> path ? where is this coming from?

was that you use `path` but it is not described in the spec in your PR.

also side note: I think it should be `pathname` as we do not want to allow search (query params) here -> https://url.spec.whatwg.org/#dom-url-pathname

Reply:
> Can we then improve the example and not put https protocol but something that will not cause brains to fart?

> or best would be to have 2 examples, one with path and one without it:

Absolutely. I'll add two examples.

> was that you use path but it is not described in the spec in your PR.

True. My fault. Adding it.

> also side note: I think it should be pathname as we do not want to allow search (query params) here -> https://url.spec.whatwg.org/#dom-url-pathname

Good catch. Changing it 👍